File: Solutions/collective/gather.f90
program gather
    use mpi
    implicit none

    integer (kind=4) :: i,ierror, world_rank, world_size
    character (len=1) :: letter
    real (kind=4) :: number(1000)
    character (len=1), allocatable :: buf(:)
  
    call MPI_Init( ierror )
    if (ierror .ne. 0) stop ' Cannot initialize MPI '

! Find our world rank & size to print it
    call MPI_Comm_rank( MPI_COMM_WORLD, world_rank, ierror )
    call MPI_Comm_size( MPI_COMM_WORLD, world_size, ierror )


    call random_seed( )
    call random_number(number)
    number = number*25.0 + 97.0;
    letter = achar(nint(number(10)))

! We need an array buf to hold one element from each process
    allocate(buf(0:world_size-1))

! Do the gather onto rank 0
    call MPI_Gather( letter, 1, MPI_CHARACTER, buf, 1, MPI_CHARACTER, 0, MPI_COMM_WORLD, ierror )


! Print the value on all processes
    if ( world_rank .eq. 0 )  then
            write(6,'(a,i0,a,i0)') 'I am process ',world_rank,' of ',world_size
        do i = 0,world_size-1
            write(6,'(a1)',advance='NO')  buf(i)
        end do
        write(6,*)
    endif

! Do the gather onto rank 1. We expect the same results
    call MPI_Gather( letter, 1, MPI_CHARACTER, buf, 1, MPI_CHARACTER, 1, MPI_COMM_WORLD, ierror )

! Print the value on all processes
    if ( world_rank .eq. 1 )  then
            write(6,'(a,i0,a,i0)') 'I am process ',world_rank,' of ',world_size
        do i = 0,world_size-1
           write(6,'(a1)',advance='NO') buf(i)
        end do
        write(6,*)
    endif

    call MPI_Finalize(ierror);
end program gather

File: Solutions/collective/gather.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

int main( int argc, char **argv ) {
    int ierror;
    ierror = MPI_Init( &argc, &argv );

    int world_rank, world_size;

    // Find our world rank & size to print it
    ierror = MPI_Comm_rank( MPI_COMM_WORLD, &world_rank );
    ierror = MPI_Comm_size( MPI_COMM_WORLD, &world_size );

    srand(time(NULL)*world_rank);
    //generates a random number between 97 and 122
    int number = rand() % 26 + 97;
    //transforms number in a character
    char letter = number;

    // We need an array buf to hold one element from each process
    char *buf;
    buf = malloc( sizeof( char ) * world_size );

    // Do the gather onto rank 0
    ierror = MPI_Gather( &letter, 1, MPI_CHAR, buf, 1, MPI_CHAR, 0, MPI_COMM_WORLD );

    int i;
    // Print the value on all processes
    if ( world_rank == 0 )  {
            printf( "I am process %d of %d\n", world_rank, world_size );
        for ( i = 0; i < world_size; i++ ) {
            printf( "%c ", buf[i] );
        }
        printf ("\n" );
    }

    // Do the gather onto rank 1. We expect the same results
    ierror = MPI_Gather( &letter, 1, MPI_CHAR, buf, 1, MPI_CHAR, 1, MPI_COMM_WORLD );

    // Print the value on all processes
    if ( world_rank == 1 )  {
            printf( "I am process %d of %d\n", world_rank, world_size );
        for ( i = 0; i < world_size; i++ ) {
            printf( "%c ", buf[i] );
        }
        printf ("\n" );
    }

    ierror = MPI_Finalize();
    return ierror;
}
